Update docs in email module (#173)
ballerina-platform/module-ballerina-email: email-ballerina/Module.md

```ballerina
email:ImapConfiguration imapConfig = {
    port: 993,
    enableSsl: true
};

email:ImapClient|email:Error imapClient = new ("imap.email.com", "reader@email.com", "pass456", imapConfig);
```

##### Receiving an email
Once the `email:ImapClient` is created, emails can be received using the IMAP4 protocol through that client.
Samples for this operation can be found below.

```ballerina
email:Message|email:Error emailResponse = imapClient->receiveMessage();
```

#### POP3 and IMAP Listeners

As POP3 and IMAP4 protocols are similar in the listener use cases, lets consider POP3.
In order to receive emails one-by-one from a POP3 server, you must first create an `email:PopListener` object.
The code for creating an `email:PopListener` can be found below.

```ballerina
listener email:PopListener emailListener = check new ({
    host: "pop.email.com",
    username: "reader@email.com",
    password: "pass456",
    pollingInterval: 2,
    port: 995
});
```

Once initialized a `service` can listen to the new emails as follows. New emails get received at the `onMessage`

> method and when errors happen `onError` method get called.

Reviewer: ```suggestion
method and when errors happen `onError` method gets called.
```

> ## Overview

Reviewer 2: What if we have a high-level section for security which discusses about SSL, SmartTLS and default behavior. Then a bit on authentication.

Reviewer 2: Also, I think we can explain a bit about the types of messages supported HTML and Text.

Reviewer 2: Lastly we can explain a bit about attachments.

Author: Added the content.

> This module contains functions to perform email operations such as sending and reading emails using the SMTP, POP3, and IMAP4 protocols.

Reviewer 3: ```suggestion
This module provides APIs to perform email operations such as sending and reading emails using the SMTP, POP3, and IMAP4 protocols.
```

Reviewer 3: make consistent with other modules

### Client

This module supports the following three client types.

- `email:SmtpClient` - The client, which supports sending an email using the SMTP protocol.
- `email:PopClient` - The client, which supports receiving an email using the POP3 protocol.
- `email:ImapClient` - The client, which supports receiving an email using the IMAP4 protocol.

#### SMTP Client

To send an email using the SMTP protocol, you must first create an `email:SmtpClient` object. The code for creating an `email:SmtpClient` can be found
 below.

##### Creating a client

The following code creates an SMTP client, which connects to the default port(465) and enables SSL.
```ballerina
email:SmtpClient smtpClient = new ("smtp.email.com",
                                   "sender@email.com",
                                   "pass123");
```
The port number of the server can be configured by passing the following configurations.

```ballerina
email:SmtpConfiguration smtpConfig = {

>     port: 465 // Can use ports, 465, 587 or 25

Reviewer 2: I think the comment is not needed.

Author: Removed.

};

> email:SmtpClient smtpClient = new ("smtp.email.com",

Reviewer 2: Is this correct ? I thought init could return error ?

Author: Yes. Added the error.

                                   "sender@email.com",
                                   "pass123",
                                    smtpConfig);

> ```

Reviewer 2: I think it could be formatted into a single line.

Author: Made it a single line.

##### Sending an email

Once the `email:SmtpClient` is created, an email can be sent using the SMTP protocol through that client.
Samples for this operation can be found below.

```ballerina
email:Message email = {
    to: ["receiver1@email.com", "receiver2@email.com"],
    cc: ["receiver3@email.com", "receiver4@email.com"],
    bcc: ["receiver5@email.com"],
    subject: "Sample Email",
    body: "This is a sample email.",
    'from: "author@email.com",
    sender: "sender@email.com",
    replyTo: ["replyTo1@email.com", "replyTo2@email.com"]
};

> email:Error? response = smtpClient->sendMessage(email);

Reviewer 2: Lets use check here ?

Author: Moved to `check` based error handling.

```

An email can be sent directly by calling the client, specifying optional parameters as named parameters, as well.
Samples for this operation can be found below.

```ballerina
email:Error? response = smtpClient->send(
    ["receiver1@email.com", "receiver2@email.com"],
    "Sample Email",
    "author@email.com",
    body="This is a sample email.",
    cc=["receiver3@email.com", "receiver4@email.com"],
    bcc=["receiver5@email.com"],
    sender="sender@email.com",
    replyTo=["replyTo1@email.com", "replyTo2@email.com"]
);
```

#### POP3 Client

To receive an email using the POP3 protocol, you must first create an `email:PopClient` object. The code for creating an
 `email:PopClient` can be found below.

##### Creating a client

The following code creates a POP3 client, which connects to the default port(995) and enables SSL.
```ballerina

> email:PopClient|email:Error popClient = new ("pop.email.com",

Reviewer 2: ditto

Author: Moved to `check` based error handling.

                                             "reader@email.com",
                                             "pass456");

> ```

Reviewer 2: I think we can stick to a single line.

Author: Made it a single line.

The port number of the server can be configured by passing the following configurations.
```ballerina
email:PopConfiguration popConfig = {
    port: 995
};

> email:PopClient|email:Error popClient = new ("pop.email.com",

Reviewer 2: ditto

Author: Moved to `check` based error handling.

                                             "reader@email.com",
                                             "pass456",
                                              popConfig);

> ```

Reviewer 2: ditto

Author: Made it a single line.

The code for creating an `email:PopListener` can be found below.

```ballerina
listener email:PopListener emailListener = check new ({
    host: "pop.email.com",
    username: "reader@email.com",
    password: "pass456",
    pollingInterval: 2,
    port: 995
});
```

Once initialized a `service` can listen to the new emails as follows. New emails get received at the `onMessage`
method and when errors happen `onError` method get called.

```ballerina
service "emailObserver" on emailListener {

    remote function onMessage(email:Message emailMessage) {
        io:println("Email Subject: ", emailMessage.subject);
        io:println("Email Body: ", emailMessage.body);
    }

    remote function onError(email:Error emailError) {
        io:println("Error while polling for the emails: " + emailError.message());
    }

}
```

For information on the operations, which you can perform with this module, see the **Functions**  below. For examples of the usage of the operation, see the following.
  * [Send Emails Example](https://ballerina.io/learn/by-example/send-email.html)
  * [Receive Emails using a client Example](https://ballerina.io/learn/by-example/receive-email-using-client.html)

>   * [Receive Emails using a listener Example](https://ballerina.io/learn/by-example/receive-email-using-listener.html)

Reviewer 2: We decided to remove BBEs from the API documentation.

Author: Removed.

```

#### POP3 and IMAP Listeners

As POP3 and IMAP4 protocols are similar in the listener use cases, lets consider POP3.
In order to receive emails one-by-one from a POP3 server, you must first create an `email:PopListener` object.
The code for creating an `email:PopListener` can be found below.

```ballerina
listener email:PopListener emailListener = check new ({
    host: "pop.email.com",
    username: "reader@email.com",
    password: "pass456",
    pollingInterval: 2,
    port: 995
});
```

Once initialized a `service` can listen to the new emails as follows. New emails get received at the `onMessage`
method and when errors happen `onError` method gets called.

```ballerina
service "emailObserver" on emailListener {

    remote function onMessage(email:Message emailMessage) {
        io:println("Email Subject: ", emailMessage.subject);
        io:println("Email Body: ", emailMessage.body);
    }

    remote function onError(email:Error emailError) {
        io:println("Error while polling for the emails: " + emailError.message());
    }

}
```

## Security and Authentication

Email module supports both TLS/SSL and STARTTLS as transport level security.

> Transport level security for all SMTP, POP3, and IMAP clients/listeners can be configured with `secureSocket` field.

Reviewer 4: ```suggestion
Transport-level security for all SMTP, POP3, and IMAP clients/listeners can be configured with the `secureSocket` field.
```

Author: Same as above.

```ballerina
secureSocket: {
    cert: "path/to/certfile.crt",
    protocol: {
        name: TLS,
        versions: ["TLSv1.2", "TLSv1.1"]
    },
    ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"],
    verifyHostName: true
}
```

> Transport level security for SMTP client configuration can be defined as follows.

Reviewer 4: ```suggestion
Transport-level security for the SMTP client configuration can be defined as follows.
```

```ballerina
email:SmtpConfiguration smtpConfig = {
    port: 465,
    secureSocket: {
        // Transport level configuration
    }
};
```

> Transport level security for POP3 client configuration can be defined as follows.

Reviewer 4: ```suggestion
Transport-level security for the POP3 client configuration can be defined as follows.
```

```ballerina
email:PopConfiguration popConfig = {
     port: 995,
     secureSocket: {
         // Transport level configuration
     }
};
```

> Transport level security for IMAP client configuration can be defined as follows.

Reviewer 4: ```suggestion
Transport-level security for the IMAP client configuration can be defined as follows.
```

```ballerina
email:ImapConfiguration imapConfig = {
     port: 993,
     secureSocket: {
         // Transport level configuration
     }
};
```

> Transport level security for POP3 listener configuration can be defined as follows.

Reviewer 4: ```suggestion
Transport-level security for the POP3 listener configuration can be defined as follows.
```

```ballerina
email:PopListenerConfiguration popListenerConfig = {
    host: "127.0.0.1",
    username: "hascode",
    password: "abcdef123",
    pollingInterval: 2,
    port: 995,
    secureSocket: {
        // Transport level configuration
    }
};
```

> Transport level security for IMAP listener configuration can be defined as follows.

Reviewer 4: ```suggestion
Transport-level security for the IMAP listener configuration can be defined as follows.
```

```ballerina
email:ImapListenerConfiguration imapListenerConfig = {
    host: "127.0.0.1",
    username: "hascode",
    password: "abcdef123",
    pollingInterval: 2,
    port: 993,
    secureSocket: {
        // Transport level configuration
    }
};
```

> By default, TLS/SSL is enabled as the default transport level security protocol and the certificate verification is set as required.

Reviewer 4: ```suggestion
By default, TLS/SSL is enabled as the default transport-level security protocol, and the certificate verification is set as required.
```

> This optional protocol definition can be configured with the `enum` field, `security`, in each of the configuration type described above.

Reviewer 4: ```suggestion
This optional protocol definition can be configured with the `security` `enum` field in each of the configuration types described above.
```

> Options available with the field, `security` are as follows.

Reviewer 4: ```suggestion
The options available with the `security` field are as follows.
```

1. `SSL` - As same as the default TLS/SSL protocol

> 2. `START_TLS_NEVER` - Disables both TLS/SSL and STARTTLS protocols and allows only the unencrypted transport level communication

Reviewer 4: ```suggestion
2. `START_TLS_NEVER` - Disables both TLS/SSL and STARTTLS protocols and allows only the unencrypted transport-level communication
```

> 3. `START_TLS_ALWAYS` - Makes it mandatory to use secure STARTTLS protocol

Reviewer 4: ```suggestion
3. `START_TLS_ALWAYS` - Makes it mandatory to use the secure STARTTLS protocol
```

> 4. `START_TLS_AUTO` - Enables STARTTLS protocol which would switch to unsecure communication mode if the secure STARTTLS mode is not available in the server

Reviewer 4: ```suggestion
4. `START_TLS_AUTO` - Enables the STARTTLS protocol, which would switch to the unsecured communication mode if the secure STARTTLS mode is not available in the server
```

> Following is an example of using the field, `security` in the SMTP client with `START_TLS_AUTO` mode.

Reviewer 4: ```suggestion
The following is an example of using the `security` field in the SMTP client with the `START_TLS_AUTO` mode.
```

```ballerina
email:SmtpConfiguration smtpConfig = {
    port: 587,
    secureSocket: {
        // Transport level configuration
    },
    security: START_TLS_AUTO
};
```

> Similarly, other client/listener configuration types can also be defined with the field `security`.

Reviewer 4: ```suggestion
Similarly, other client/listener configuration types can also be defined with the `security` field.
```

**Note**: Make sure the port number is changed accordingly depending on the protocol used.

Standard port numbers used for each of the protocol for each type of transport security are as given below.

| Protocol/Security | SSL | STARTTLS | Unsecure |
|-------------------|-----|----------|----------|
| **SMTP**          | 465 | 587      | 25, 587  |
| **POP3**          | 995 | 995      | 110      |
| **IMAP4**         | 993 | 143, 993 | 143      |

> All the authentication is based on username/password credentials.

Reviewer 4: ```suggestion
All the authentications are based on the username/password credentials.
```

> Note that when the `'from` field is not provided in an `email:Message`, the `username` field of the initialization argument to the `email:SmtpClient` is set as the `from` address of an email to be sent with SMTP.

Reviewer 4: ```suggestion
>**Note:** When the `'from` field is not provided in an `email:Message`, the `username` field of the initialization argument of the `email:SmtpClient` is set as the `from` address of an email to be sent with SMTP.
```

## Message Content and Attachments

> An `email:Message` prepared to be sent, can have the text body content, `body`, and/or HTML body content, `htmlBody`.

Reviewer 4: ```suggestion
An `email:Message` prepared to be sent can have the text body content, `body`, and/or HTML body content (`htmlBody`).
```

> When emails are received with POP3 or IMAP, text email bodies and HTML bodies of the email are captured by `body` and `htmlBody` fields of `email:Message` respectively.

Reviewer 4: ```suggestion
When emails are received with POP3 or IMAP, the text email bodies and HTML bodies of the email are captured by the `body` and `htmlBody` fields of the `email:Message` respectively.
```

> When sending emails with SMTP, there are four options to specify email `attachments` in `email:Message`.

Reviewer 4: ```suggestion
When sending emails with SMTP, there are four options to specify the email `attachments` in the `email:Message`.
```

> 1. With `email:Attachment` type which points to an attachment file along with its content-type

Reviewer 4: ```suggestion
1. With the `email:Attachment` type, which points to an attachment file along with its content-type
```

> 2. With an array of `email:Attachment` type

Reviewer 4: ```suggestion
2. With an array of the `email:Attachment` type
```

> 3. With `mime:Entity` type

Reviewer 4: ```suggestion
3. With the `mime:Entity` type
```

> 4. With an array of `mime:Entity` type

Reviewer 4: ```suggestion
4. With an array of the `mime:Entity` type
```

> Option 1 and 2 are designed for ordinary users to attach files from the local machine along with its content-type.

Reviewer 4: ```suggestion
Option 1 and 2 are designed for ordinary use to attach files from the local machine along with its content-type.
```

Author: Actually meant about users.

Option 3 and 4 are designed for advanced users who have programming knowledge to define complex MIME typed data attachments.

> Following is an example of attaching a PDF file to an email with option 1.

Reviewer 4: ```suggestion
The following is an example of attaching a PDF file to an email with option 1.
```

Author: Mistakenly skipped committing. Then I added the fix again below.

```ballerina
email:Attachment pdfAttachment = {filePath: "path/to/application.pdf", contentType: "application/pdf"};

email:Message email = {
    // Other fields
    attachments: pdfAttachment
};
```

> Following is an example of attaching a JPG file to an email with option 3.

Reviewer 4: ```suggestion
The following is an example of attaching a JPG file to an email with option 3.
```

> Following is an example of attaching a JPG file to an email with option 3.

Author: ```suggestion
The following is an example of attaching a JPG file to an email with option 3.
```
